File: solutions/misc/frequently-asked-questions/sliding-window-problems.py
"""
Sliding Window — 5 Core Problems (Python)

LC 76  — Minimum Window Substring
LC 209 — Minimum Size Subarray Sum
LC 424 — Longest Repeating Character Replacement
LC 567 — Permutation in String
LC 904 — Fruit Into Baskets
"""

from collections import Counter, defaultdict
from typing import List


# LC 76 — Minimum Window Substring — O(|s|+|t|)
def min_window(s: str, t: str) -> str:
    if len(t) > len(s):
        return ""
    # Count of each character needed from t
    need = Counter(t)
    # Count of each character in current window
    window: Counter = Counter()
    # have = number of unique chars in window that meet required count
    # required = number of unique chars in t
    have, required = 0, len(need)
    left, res, res_len = 0, "", float("inf")

    for right, c in enumerate(s):
        window[c] += 1
        # If current char is in t and we just met its required count, increment have
        if c in need and window[c] == need[c]:
            have += 1
        # When window contains all required chars, try to shrink from left
        while have == required:
            if right - left + 1 < res_len:
                res_len = right - left + 1
                res = s[left:right+1]
            # Remove leftmost char from window and adjust have if needed
            lc = s[left]; left += 1
            window[lc] -= 1
            if lc in need and window[lc] < need[lc]:
                have -= 1
    return res


# LC 209 — Minimum Size Subarray Sum — O(n)
def min_sub_array_len(target: int, nums: List[int]) -> int:
    left = total = 0
    res = float("inf")
    for right, n in enumerate(nums):
        total += n
        # Shrink window from left while sum >= target
        while total >= target:
            res = min(res, right - left + 1)
            total -= nums[left]; left += 1
    return 0 if res == float("inf") else res


# LC 424 — Longest Repeating Character Replacement — O(n)
def character_replacement(s: str, k: int) -> int:
    count = [0] * 26
    left = max_freq = res = 0
    for right, c in enumerate(s):
        count[ord(c) - 65] += 1
        max_freq = max(max_freq, count[ord(c) - 65])
        # If window size - max_freq > k, we need more than k replacements → shrink
        while right - left + 1 - max_freq > k:
            count[ord(s[left]) - 65] -= 1; left += 1
        res = max(res, right - left + 1)
    return res


# LC 567 — Permutation in String — O(n)
def check_inclusion(s1: str, s2: str) -> bool:
    if len(s1) > len(s2):
        return False
    # Frequency map of s1 (what we need to match)
    need = Counter(s1)
    # Frequency map of current window in s2
    window: Counter = Counter()
    matches = 0
    required = len(s1)
    left = 0

    for right, c in enumerate(s2):
        window[c] += 1
        # If adding this char helps match s1's frequency, increment matches
        if need[c] > 0 and window[c] <= need[c]:
            matches += 1
        # If window exceeds required size, shrink from left
        if right - left + 1 > required:
            lc = s2[left]; left += 1
            if need[lc] > 0 and window[lc] <= need[lc]:
                matches -= 1
            window[lc] -= 1
        # If all characters match, s1's permutation exists in s2
        if matches == required:
            return True
    return False


# LC 904 — Fruit Into Baskets — O(n)
# Equivalent to: longest subarray with at most 2 distinct values
def total_fruit(fruits: List[int]) -> int:
    basket: Counter = Counter()
    left = res = 0
    for right, f in enumerate(fruits):
        basket[f] += 1
        # If more than 2 fruit types, shrink window from left
        while len(basket) > 2:
            basket[fruits[left]] -= 1
            if basket[fruits[left]] == 0:
                del basket[fruits[left]]
            left += 1
        res = max(res, right - left + 1)
    return res


# ── Tests ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print(min_window("ADOBECODEBANC", "ABC"))      # "BANC"
    print(min_sub_array_len(7, [2,3,1,2,4,3]))     # 2
    print(character_replacement("AABABBA", 1))     # 4
    print(check_inclusion("ab", "eidbaooo"))       # True
    print(total_fruit([1,2,1,2,3]))                # 4

File: solutions/misc/frequently-asked-questions/sliding-window-problems.js
/**
 * Sliding Window — 5 Core Problems
 *
 * Template:
 *   left = 0
 *   for right in range(n):
 *     expand window (add arr[right])
 *     while window_invalid:
 *       shrink window (remove arr[left++])
 *     update answer
 *
 * Problems:
 *   LC 76  — Minimum Window Substring (Hard)
 *   LC 209 — Minimum Size Subarray Sum (Medium)
 *   LC 424 — Longest Repeating Character Replacement (Medium)
 *   LC 567 — Permutation in String (Medium)
 *   LC 904 — Fruit Into Baskets (Medium)
 */

// ═══════════════════════════════════════════════════════
// LC 76 — Minimum Window Substring
// Time: O(|s| + |t|)  Space: O(|t|)
// ═══════════════════════════════════════════════════════
function minWindow(s, t) {
  if (t.length > s.length) return "";
  const need = new Map();
  for (const c of t) need.set(c, (need.get(c) || 0) + 1);

  let have = 0, required = need.size;
  const window = new Map();
  let res = "", resLen = Infinity;
  let left = 0;

  for (let right = 0; right < s.length; right++) {
    const c = s[right];
    window.set(c, (window.get(c) || 0) + 1);
    if (need.has(c) && window.get(c) === need.get(c)) have++;

    while (have === required) {
      if (right - left + 1 < resLen) {
        resLen = right - left + 1;
        res = s.slice(left, right + 1);
      }
      const lc = s[left++];
      window.set(lc, window.get(lc) - 1);
      if (need.has(lc) && window.get(lc) < need.get(lc)) have--;
    }
  }
  return res;
}

// ═══════════════════════════════════════════════════════
// LC 209 — Minimum Size Subarray Sum
// Time: O(n)  Space: O(1)
// ═══════════════════════════════════════════════════════
function minSubArrayLen(target, nums) {
  let left = 0, sum = 0, minLen = Infinity;
  for (let right = 0; right < nums.length; right++) {
    sum += nums[right];
    while (sum >= target) {
      minLen = Math.min(minLen, right - left + 1);
      sum -= nums[left++];
    }
  }
  return minLen === Infinity ? 0 : minLen;
}

// ═══════════════════════════════════════════════════════
// LC 424 — Longest Repeating Character Replacement
// Time: O(n)  Space: O(26)
// ═══════════════════════════════════════════════════════
function characterReplacement(s, k) {
  const count = new Array(26).fill(0);
  let left = 0, maxFreq = 0, res = 0;

  for (let right = 0; right < s.length; right++) {
    count[s.charCodeAt(right) - 65]++;
    maxFreq = Math.max(maxFreq, count[s.charCodeAt(right) - 65]);

    // window size - maxFreq = replacements needed
    while (right - left + 1 - maxFreq > k) {
      count[s.charCodeAt(left++) - 65]--;
    }
    res = Math.max(res, right - left + 1);
  }
  return res;
}

// ═══════════════════════════════════════════════════════
// LC 567 — Permutation in String
// Time: O(n)  Space: O(26)
// ═══════════════════════════════════════════════════════
function checkInclusion(s1, s2) {
  if (s1.length > s2.length) return false;
  const need = new Array(26).fill(0);
  const have = new Array(26).fill(0);

  for (const c of s1) need[c.charCodeAt(0) - 97]++;

  let left = 0, matches = 0;
  const required = s1.length;

  for (let right = 0; right < s2.length; right++) {
    const idx = s2.charCodeAt(right) - 97;
    have[idx]++;
    if (need[idx] > 0 && have[idx] <= need[idx]) matches++;

    if (right - left + 1 > required) {
      const li = s2.charCodeAt(left++) - 97;
      if (need[li] > 0 && have[li] <= need[li]) matches--;
      have[li]--;
    }

    if (matches === required) return true;
  }
  return false;
}

// ═══════════════════════════════════════════════════════
// LC 904 — Fruit Into Baskets
// = Longest subarray with at most 2 distinct values
// Time: O(n)  Space: O(1) — at most 2 types
// ═══════════════════════════════════════════════════════
function totalFruit(fruits) {
  const basket = new Map();
  let left = 0, res = 0;

  for (let right = 0; right < fruits.length; right++) {
    basket.set(fruits[right], (basket.get(fruits[right]) || 0) + 1);

    while (basket.size > 2) {
      const lf = fruits[left++];
      basket.set(lf, basket.get(lf) - 1);
      if (basket.get(lf) === 0) basket.delete(lf);
    }
    res = Math.max(res, right - left + 1);
  }
  return res;
}

// ─── Tests ───────────────────────────────────────────────────────────────────
console.log(minWindow("ADOBECODEBANC", "ABC"));           // "BANC"
console.log(minSubArrayLen(7, [2,3,1,2,4,3]));            // 2
console.log(characterReplacement("AABABBA", 1));          // 4
console.log(checkInclusion("ab", "eidbaooo"));            // true
console.log(totalFruit([1,2,1,2,3]));                     // 4

module.exports = { minWindow, minSubArrayLen, characterReplacement, checkInclusion, totalFruit };
